mentorship iosTests/MentorshipTests.swift

```diff
@@ -7,26 +7,80 @@
 import XCTest
 @testable import mentorship_ios
 
-class MentorshipTests: XCTestCase {
+class MockURLProtocol: URLProtocol {
+    static var requestHandler: ((URLRequest) throws -> (HTTPURLResponse, Data))?
+    
+    override class func canInit(with request: URLRequest) -> Bool {
+        return true
+    }
+    
+    override class func canonicalRequest(for request: URLRequest) -> URLRequest {
+        return request
+    }
+    
+    override func startLoading() {
+        guard let handler = MockURLProtocol.requestHandler else {
+            XCTFail("Received unexpected request with no handler set")
+            return
+        }
+        do {
+            let (response, data) = try handler(request)
+            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
+            client?.urlProtocol(self, didLoad: data)
+            client?.urlProtocolDidFinishLoading(self)
+        } catch {
+            client?.urlProtocol(self, didFailWithError: error)
+        }
+    }
+    
+    override func stopLoading() {
+    }
+}
 
+
+class MentorshipTests: XCTestCase {
+    var urlSession: URLSession!
+    
     override func setUpWithError() throws {
-        // Put setup code here. This method is called before the invocation of each test method in the class.
+        // Set url session for mock networking
+        let configuration = URLSessionConfiguration.ephemeral
+        configuration.protocolClasses = [MockURLProtocol.self]
+        urlSession = URLSession(configuration: configuration)
     }
-
+    
     override func tearDownWithError() throws {
         // Put teardown code here. This method is called after the invocation of each test method in the class.
     }
-
-    func testExample() throws {
-        // This is an example of a functional test case.
-        // Use XCTAssert and related functions to verify your tests produce the correct results.
+    
+    func testRequestAction() throws {
+        // Set mock json for server response
+        let mockJSON = ResponseMessage(message: "response message")
+        // Create mock data from mock json
+        let mockData = try JSONEncoder().encode(mockJSON)
+        
+        // Use mock protocol, and return mock data and url response from handler
+        MockURLProtocol.requestHandler = { request in
+            return (HTTPURLResponse(), mockData)
+        }
+        
+        // Init API Class with mock testing session
+        let requestActionAPI = RelationRequestActionAPI(urlSession: urlSession)
+        // Init expectation (used to test async code)
+        let expectation = XCTestExpectation(description: "response")
+        
+        // mock api call and test in handler
+        requestActionAPI.actOnPendingRequest(action: .accept, reqID: 0) {
+            XCTAssertEqual($0.message, mockJSON.message)
```

Comment:
are you actually testing here network request? I think we spoke we do not test server, we mock the response for view model test purpose.

Reply:
No. By using the MockURLProtocol and a custom urlsession for that, we are able to mock network calls. And now, we don't unnecessary protocols for network apis, it's pretty neat! I think, this resolves the previous comment too.

Reply:
what if your data is not coming from network call?

Reply:
For that, we can just make another function inside this class?

Reply:
This is not a way you keep isolation between layers. Adding new data source should not modify either view or view tests.

Reply:
I don't totally understand this. Can you elaborate on it further?

Reply:
yes we do not test data source, because data source are having own tests to make sure they work. We test local logic, local parsing, local UI. Does it makes sense?

Reply:
Okay. The completion handler is to be defined in the view right? So, are we basically trying to give mockData and then test that handler?